astrolabe/runner.py

```diff
@@ -163,34 +173,73 @@ def run(self, persist_cluster=False, startup_time=1):
-
-        # Sleep before polling to give Atlas time to update cluster.stateName.
-        sleep(3)
-
-        # Step-4: wait until maintenance completes (cluster is IDLE).
-        selector = BooleanCallablePoller(
-            frequency=self.config.polling_frequency,
-            timeout=self.config.polling_timeout)
-        LOGGER.info("Waiting for cluster maintenance to complete")
-        selector.poll([self], attribute="is_cluster_state", args=("IDLE",),
-                      kwargs={})
-        self.verify_cluster_configuration_matches("final")
-        LOGGER.info("Cluster maintenance complete")
+        for operation in self.spec.operations:
+            if len(operation) != 1:
+                raise ValueError("Operation must have exactly one key: %s" % operation)
+                
+            op_name, op_spec = next(iteritems(operation))
+            
+            if op_name == 'setClusterConfiguration':
+                # Step-3: begin maintenance routine.
+                final_config = op_spec
+                cluster_config = final_config.clusterConfiguration
+                process_args = final_config.processArgs
+
+                if not cluster_config and not process_args:
+                    raise RuntimeError("invalid maintenance plan")
+
+                if cluster_config:
+                    LOGGER.info("Pushing cluster configuration update")
+                    self.cluster_url.patch(**cluster_config)
+
+                if process_args:
+                    LOGGER.info("Pushing process arguments update")
+                    self.cluster_url.processArgs.patch(**process_args)
+
+                # Step-4: wait until maintenance completes (cluster is IDLE).
+                self.wait_for_idle()
+                self.verify_cluster_configuration_matches(final_config)
+                LOGGER.info("Cluster maintenance complete")
+                
+            if op_name == 'testFailover':
+                self.cluster_url['restartPrimaries'].post()
+                
+                self.wait_for_idle()
+                
+            if op_name == 'sleep':
+                _time.sleep(op_spec)
+                
+            if op_name == 'waitForIdle':
+                self.wait_for_idle()
+                
+            if op_name == 'restartVms':
+                url = "/api/private/nds/groups/%s/clusters/%s/reboot" % (self.project.id, self.cluster_name)
+                self.admin_client.request('POST', url)
+                
+                self.wait_for_idle()
+                
+            if op_name == 'assertPrimaryRegion':
+                region = op_spec['region']
+                
+                cluster_config = self.cluster_url.get().data
+                timer = Timer()
+                timer.start()
+                timeout = op_spec.get('timeout', 90)
+                
+                while True:
+                    with mongo_client(self.get_connection_string()) as mc:
```

Comment:
Why are we not simply using `self.client` here? We don't seem to modify any of the URI options and directly use the connection string so creating a new client everytime to run this operation appears to be pointless. Can you add the rationale for using a new client as a comment?

Reply:
As I understand it `self.client` is an `AtlasClient`, not a `MongoClient` hence wouldn't do the job.

Reply:
Looking at this again I think one client instance for all iterations will suffice therefore I'll lift client instantiation to be outside of the loop.